src/crush/CrushWrapper.cc

```diff
@@ -1817,6 +1821,144 @@ int CrushWrapper::update_device_class(int id,
   return 1;
 }
 
+// Recursively get all only-child shadow parents up to the root
+// based on the given item id.
+// Return -EBUSY if any parent has more than one children.
+// Return -ENOENT if a specific parent can not be found.
+// Return 0 on success.
+int CrushWrapper::get_immediate_only_child_shadow_parents(CephContext *cct,
+                                                          int id,
+                                                          list<int> *parents)
+{
+  assert(parents);
+  parents->clear();
+  while (_search_item_exists(id)) {
```

Comment:
> we aren't modifying the tree inside the loop

`id ` will be reset to its direct parent after every loop, see https://github.com/ceph/ceph/pull/16388/commits/b80c5302c88d9319424d4bda7da71da89e32ff39#diff-c1ea4e8ae4b5520080ef65c1bd8ed666R1857
So we'll quit(normally) if we finally reach the root.